Update Block validation to support error context

martynmjones opened the pull request:
### Changes proposed in this Pull Request:

Closes #2507 

Updates Block validation to return an object containing the error message and the clientId as context to support the new link displayed in the snackbar.

<img width="983" alt="Screenshot 2024-08-07 at 16 09 11" src="https://github.com/user-attachments/assets/e38ce68c-893c-4241-bf7c-727d814afee5">

Additionally, with WooCommerce `9.2.0-rc.1` the `Multipack` field validation changes so if a decimal value was entered a blank error message was displayed:

<img width="1002" alt="Screenshot 2024-08-07 at 21 58 27" src="https://github.com/user-attachments/assets/f6303264-5f77-49cb-9842-19d8e3351bf4">

As it's not possible to either set a default validation error message or add custom rules outside of the supported few, the field has been updated to use a text field instead of number. A [pattern is then supplied with a custom error message](https://github.com/woocommerce/google-listings-and-ads/blob/681a07c3c107ccc33014680fa9ad64a3e3812737/src/Admin/Input/Integer.php#L23) to prevent users from entering non numeric values, decimal values, or negative values.

<img width="1071" alt="Screenshot 2024-08-08 at 12 06 32" src="https://github.com/user-attachments/assets/78c10230-01b2-4eff-bb06-fb873473d45a">

### Detailed test instructions:
1. `npm run wp-env:up`
1. `npm run -- wp-env run tests-cli -- wp plugin update woocommerce --version=9.2.0-rc.1`
1. `npm run -- wp-env run tests-cli -- wp wc update`
1. `npm run test:e2e`
1. Confirm tests pass
1. Enable the new product editor in `WooCommerce > Settings > Advanced > Features`
1. Edit a product and test different values in the `Multipack` field under the `Google for WooCommerce` tab
1. Confirm only whole numbers are accepted input

### Additional
⚠️  Please merge once approved as this PR is required before GLA can be released with the `9.2.0` compat bump.

### Changelog entry

> Update - Block validation to support error context

martynmjones committed:
Update validation message for WooCommerce 9.2.0

martynmjones committed:
Update Product date and time fields tests

martynmjones committed:
Use pattern for validation with text input instead of number input

martynmjones committed:
Merge remote-tracking branch 'origin/develop' into fix/2507-snackbar-error-message

martynmjones committed:
Fix linting errors

martynmjones committed:
PHPCS fix

martynmjones committed:
Fix integer and multipack input unit tests

martynmjones committed:
Unit tests PHPCS fixes

martynmjones committed:
Update field count expectations in e2e tests

martynmjones committed:
JS Linting fix

eason9487 commented on js/src/blocks/product-date-time-field/edit.js:
```diff
@@ -21,11 +22,18 @@ import styles from './editor.module.scss';
 
 	if ( ! input.validity.valid ) {
-		return input.validationMessage;
+		if ( isWcVersion( '9.2.0', '>' ) ) {
```
Looks like the comparison should be `<`, and suggest adding a `compatibility-code` comment.

```suggestion
		// compatibility-code "WC < 9.2"
		if ( isWcVersion( '9.2.0', '<' ) ) {
```

Otherwise, with WC < 9.2, it will cause an error after entering an invalid value.

https://github.com/user-attachments/assets/6adddec1-0479-423c-8db7-4ffdd69cd9b7

eason9487 commented on tests/e2e/utils/product-editor.js:
```diff
@@ -577,17 +577,15 @@ export function getProductBlockEditorUtils( page ) {
+			);
 
 			await expect( failureNotice ).toBeVisible();
+			await expect( failureNotice ).toHaveText( message );
```
This assertion failed with WC < 9.2. How about making it also compatible?

```suggestion
			await expect( failureNotice ).toContainText( message );
```

jorgemd24 replied:
Thanks for the suggestion I've updated it here: [51e3679](https://github.com/woocommerce/google-listings-and-ads/pull/2511/commits/51e36797ff3f70569e16539409aed4990bd5120d)

eason9487 commented on tests/e2e/utils/product-editor.js:
```diff
@@ -577,17 +577,15 @@ export function getProductBlockEditorUtils( page ) {
-			// Dismiss the notice.
-			await failureNoticeDismissButton.click();
+			// // Dismiss the notice.
+			await failureNotice.click();
```
This action failed with WC < 9.2. How about making it also compatible?

```suggestion
			// Dismiss the notice.
			await failureNotice
				.getByRole( 'button' ) // compatibility-code "WC < 9.2" -- Dismiss by its inner close button.
				.or( failureNotice )
				.click();

```

jorgemd24 replied:
I tried using the or method, but if both elements are present, it throws a "strict mode violation." Using nth to select one element worked for some tests but not for others. So, I decided to take a different approach, but you can adjust it if needed. [51e3679](https://github.com/woocommerce/google-listings-and-ads/pull/2511/commits/51e36797ff3f70569e16539409aed4990bd5120d)

eason9487 replied:
> I tried using the or method, but if both elements are present, it throws a "strict mode violation."

I later realized it had to be changed to this but I missed updating to the comment before submitting review.

```js
// Dismiss the notice.
await failureNotice
	.or( failureNotice.getByRole( 'button' ) )
	.last()
	.click();
```

> So, I decided to take a different approach, but you can adjust it if needed. [51e3679](https://github.com/woocommerce/google-listings-and-ads/pull/2511/commits/51e36797ff3f70569e16539409aed4990bd5120d)

Your method is better and clearer, thanks!

eason9487 commented on src/Admin/Input/Integer.php:
```diff
@@ -20,6 +20,15 @@ public function __construct() {
+		$this->set_block_attribute(
+			'pattern',
+			[
+				'value'   => '\d+',
```
Maybe the pattern could be `[0-9]\d?` to avoid saving a value like '000123'.

jorgemd24 replied:
Not sure about this. If I use `[0-9]\d`, it will only match two digits and won't work for single-digit numbers. If I switch to `[0-9]\d+`, it will match any number of digits but won't prevent cases like `000123`. To handle that, I think the pattern needs to be something like `^(0|[1-9]\d*)$`. WDYT?

eason9487 replied:
Sorry, it should be `[0-9]\d*`.

> To handle that, I think the pattern needs to be something like `^(0|[1-9]\d*)$`. WDYT?

`^(?:` and `)$` impliedly wrap up the pattern of `<input>`, so it will match the entire value rather than matching a substring.

jorgemd24 replied:
mmm, I tried using `[0-9]\d*` but I got a similar result to `0-9]\d+`, we are still able to type `00123`.

eason9487 replied:
Sorry again 😂. I just tested with `\d|[1-9]\d*` and hope it's correct this time.

eason9487 replied:
I was too focused on the `^` and `$` parts. Just noticed the pattern `0|[1-9]\d*` you commented in https://github.com/woocommerce/google-listings-and-ads/pull/2511#discussion_r1715375981 should work as well.

eason9487 replied:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/2511/commits/ea781f60c49d9b37ec68c3d4e98b1ad6cf56085e with more assertions for invalid values.

eason9487 commented on src/Admin/Input/Integer.php:
```diff
@@ -20,6 +20,15 @@ public function __construct() {
+			'pattern',
+			[
+				'value'   => '\d+',
+				// This error message is used in core so to avoid needing to translate
+				// the string a second time WooCommerce is supplied as the text domain.
+				// phpcs:ignore WordPress.WP.I18n.TextDomainMismatch
+				'message' => __( 'Please enter a valid value.', 'woocommerce' ),
```
I couldn't find the same text in WooCommerce.

Is it supposed to be "Invalid value for the field."? Or perhaps leave it unspecified and let it fall back to the default error message in WooCommerce.

jorgemd24 replied:
I couldn't find that text either, so I removed the message and switched to the fallback.: [e40b122](https://github.com/woocommerce/google-listings-and-ads/pull/2511/commits/e40b1225afb555a74e8a959de3cd008c578824fc)

eason9487 reviewed (commented):
Thanks for fixing the compatibility!

There is a blocker that the comparison parameter of calling `isWcVersion` may not be correct. Other comments are optional.

jorgemd24 committed:
Fallback to the default error message

jorgemd24 committed:
Fix the comparison command

jorgemd24 committed:
Adjust e2e tests

jorgemd24 committed:
Adjust phpunit tests

eason9487 committed:
Adjust the pattern in `src/Admin/Input/Integer.php` to better match a non-negative integer.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2511#discussion_r1714946853

eason9487 reviewed (approved):
The E2E tests related to product editor integration passed for the latest change. 

E2E test result with:
- WC 9.2.0-rc.1: https://github.com/woocommerce/google-listings-and-ads/actions/runs/10380513588/job/28740467547
- WC 9.1.4: https://github.com/woocommerce/google-listings-and-ads/actions/runs/10380933469/job/28742220894
   - Not sure why a few gtag-events tests keep failing but it should be irrelavant.
